java/main/jglome/Glome.java

```diff
@@ -0,0 +1,363 @@
+   */
+  public byte[] generateTag(byte[] msg, int cnt) throws CounterOutOfBoundsException {
+    return generateTag(msg, cnt, this.userMacKey);
+  }
+
+  /**
+   * Checks whether the peer's tag matches received message and some counter.
+   *
+   * @param peerTag tag from a peer.
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the peer to the user.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than {@code MAX_TAG_LENGTH}) or it's not equal to the prefix of a
+   * correct tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          "The received tag has invalid length. Expected " + minPeerTagLength + " or more, got "
+              + peerTag.length + ".");
+    }
+
+    byte[] truncatedTag = Arrays.copyOf(generateTag(msg, cnt, this.peerMacKey), peerTag.length);
+    if (!Arrays.equals(peerTag, truncatedTag)) {
+      throw new WrongTagException("The received tag doesn't match the expected tag.");
+    }
+  }
+
+  /**
+   * Common part of the constructors. Initialises peer's public key, minimum length of a peer's tag,
+   * base point for Curve25519 curve. Calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum length of a peer's tag.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  private void initMacKeys(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    this.peerKey = peerKey;
+    if (minPeerTagLength < MIN_TAG_LENGTH || minPeerTagLength > MAX_TAG_LENGTH) {
+      throw new MinPeerTagLengthOutOfBoundsException(
+          "minPeerTagLength argument should be in [1..32] range. Got " + minPeerTagLength + ".");
+    }
+    this.minPeerTagLength = minPeerTagLength;
+
+    KeyAgreement ka;
+    try {
+      ka = KeyAgreement.getInstance("XDH");
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    try {
+      ka.init(userPrivateKey());
+      ka.doPhase(peerKey, true);
+    } catch (InvalidKeyException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    byte[] sharedSecret = ka.generateSecret();
+    byte[] userMac = reverseByteArray(
+        getU32Bytes(userPublicKey())); // in order to have <i>little-endian</i> byte-order;
+    byte[] peerMac = reverseByteArray(getU32Bytes(peerKey)); // the same
+
+    this.userMacKey = getMacKey(sharedSecret, peerMac, userMac);
+    this.peerMacKey = getMacKey(sharedSecret, userMac, peerMac);
+  }
+
+  private byte[] getU32Bytes(XECPublicKey pk) {
```

Comment:
Have you considered retrieving the byte array from the encoding instead? `PublicKey.getEncoded()` returns something that can probably be parsed into X.509, and then you should be able to read the public key in correct form from the ASN.1 (at least if https://tools.ietf.org/html/rfc8410#section-4 is to be trusted here).

Reply:
Haven't look close to it, thank you. Will fix.